src/common/cockpitmemfdread.c

```diff
@@ -0,0 +1,118 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2020 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#define _GNU_SOURCE
+
+#include "cockpitmemfdread.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+static gchar *
+cockpit_memfd_read (int      fd,
+                    GError **error)
+{
+  int seals = fcntl (fd, F_GET_SEALS);
+  if (seals == -1)
+    {
+      g_set_error (error, G_FILE_ERROR, g_file_error_from_errno (errno),
+                   "could not query seals on fd %d: not memfd?: %m", fd);
+      return NULL;
+    }
+
+  const guint expected_seals = F_SEAL_WRITE | F_SEAL_GROW | F_SEAL_SHRINK;
+  if ((seals & expected_seals) != expected_seals)
+    {
+      g_set_error (error, G_FILE_ERROR, G_FILE_ERROR_INVAL,
+                   "memfd fd %d has incorrect seals set: %u (instead of %u)\n",
+                   fd, seals & expected_seals, expected_seals);
+      return NULL;
+    }
+
+  struct stat buf;
+  if (fstat (fd, &buf) != 0)
+    {
+      g_set_error (error, G_FILE_ERROR, g_file_error_from_errno (errno),
+                   "Failed to stat memfd %d: %m", fd);
+      return NULL;
+    }
+
+  if (buf.st_size < 1)
+    {
+      g_set_error (error, G_FILE_ERROR, G_FILE_ERROR_INVAL,
+                   "memfd %d must not be empty", fd);
+      return NULL;
+    }
+
+  g_autofree gchar *content = g_malloc (buf.st_size + 1);
+  gssize s = pread (fd, content, buf.st_size + 1, 0);
+  if (s == -1)
+    {
+      g_set_error (error, G_FILE_ERROR, g_file_error_from_errno (errno),
+                   "failed to read memfd %d: %m", fd);
+      return NULL;
+    }
+  else if (s != buf.st_size)
+    {
+      g_set_error (error, G_FILE_ERROR, G_FILE_ERROR_INVAL,
+                   "memfd %d changed size from %zu to %zu bytes", fd, (gssize) buf.st_size, s);
```

Comment:
Is that really because it changed (which seems a bit unlikely), or can pread() have short reads as well? I surely hope that this Does Not Happen™ on a memfd, though :grin: The original code's warning said "incomplete read", too.

Reply:
I did your trick of reading the extra byte at the end of the buffer to make sure we don't get more than expected.

Honestly, I can't imagine that this will ever fail: we're talking about something that we just checked for being sealed.  It's really a case of extreme paranoia.